### PR TITLE
[function](date function)add new date function 'last_day'

### LINKED ISSUE #13139: [Enhancement] Add new builtin function of the last day of month
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

Add a new builtin function named **last_day**, the mechanism of this function is the same as **last_day** in mysql
### Syntax
last_day(datetime)
### Arguments
**datetime** — This argument's type could be date, datetime, datev2 or datetimev2
### Returned values
This function return **date** when argument is date or datetime; return **datev2** when argument is datev2 or datetimev2.
And this function always return the month last day, so the output day may be '28', '29', '30', '31', it depends on the month.
Input ('2021-12-01'), it returns ('2021-12-31'); or input ('2000-02-05'), it returns ('2000-02-29')

### Solution

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by lsy3993**
# Proposed changes

Issue Number: close #13139

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [x] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/exprs/timestamp_functions.cpp [hunk 1]
```diff
@@ -707,6 +707,38 @@ DateTimeVal TimestampFunctions::from_days(FunctionContext* ctx, const IntVal& da
...
 }
 
+DateTimeVal TimestampFunctions::last_day(FunctionContext* ctx, const DateTimeVal& ts_val) {
+    if (ts_val.is_null) {
+        return DateTimeVal::null();
+    }
+
+    DateTimeValue ts_value = DateTimeValue::from_datetime_val(ts_val);
+
+    bool is_leap_year = (ts_value.year() % 400 == 0) ||
```
Comment (stalary) [inline be/src/exprs/timestamp_functions.cpp:717]: You can use doris::is_leap(year)

File: be/src/vec/functions/function_timestamp.cpp [hunk 2]
```diff
@@ -569,6 +569,164 @@ class FunctionUnixTimestamp : public IFunction {
...
+                    continue;
+                }
+                UInt64 cast_value = binary_cast<DateValueType, UInt64>(ts_value);
+                DataTypeDateTimeV2::cast_to_date_v2(cast_value, res_data[i]);
+            }
+        }
+    }
+
+    static int get_last_month_day(int year, int month) {
+        bool is_leap_year = (year % 400 == 0) || (year % 4 == 0 && year % 100 != 0);
```
Comment (stalary) [inline be/src/vec/functions/function_timestamp.cpp:716]: Use doris::is_leap

File: be/src/exprs/timestamp_functions.cpp (hunk 1 above)
Comment (stalary) [reply be/src/exprs/timestamp_functions.cpp:717]: Same as get_last_month_day?

File: be/src/vec/functions/function_timestamp.cpp (hunk 2 above)
Comment (lsy3993) [reply be/src/vec/functions/function_timestamp.cpp]: done

File: be/src/exprs/timestamp_functions.cpp (hunk 1 above)
Comment (lsy3993) [reply be/src/exprs/timestamp_functions.cpp]: done

File: regression-test/suites/query_p0/sql_functions/datetime_functions/test_date_function.groovy [hunk 3]
```diff
@@ -424,4 +424,30 @@ suite("test_date_function") {
     qt_sql """ select minutes_sub(test_time2,1) result from ${tableName}; """
     //seconds_sub
     qt_sql """ select seconds_sub(test_time2,1) result from ${tableName}; """
+
+    // test last_day
+    sql """ SET enable_vectorized_engine = TRUE; """
```
Comment (Gabriel39) [inline regression-test/suites/query_p0/sql_functions/datetime_functions/test_date_function.groovy:429]: also add a case for row-based engine

File: be/src/vec/functions/function_timestamp.cpp [hunk 4]
```diff
@@ -569,6 +569,164 @@ class FunctionUnixTimestamp : public IFunction {
...
+class FunctionDateOrDateTimeToDate : public IFunction {
+public:
+    static constexpr auto name = Impl::name;
+    static FunctionPtr create() {
+        return std::make_shared<FunctionDateOrDateTimeToDate<Impl, DateType>>();
+    }
+
+    String get_name() const override { return name; }
+
+    bool use_default_implementation_for_nulls() const override { return true; }
```
Comment (Gabriel39) [inline be/src/vec/functions/function_timestamp.cpp:582]: `use_default_implementation_for_constants` also need to override and set to true

File: be/src/vec/functions/function_timestamp.cpp [hunk 5]
```diff
@@ -569,6 +569,164 @@ class FunctionUnixTimestamp : public IFunction {
...
+            auto data_col = assert_cast<const ColumnVector<UInt64>*>(argument_column.get());
+            res_column = ColumnVector<UInt32>::create(input_rows_count);
+            execute_straight<DateV2Value<DateTimeV2ValueType>, UInt32, UInt64>(
+                    input_rows_count, null_map->get_data(), data_col->get_data(),
+                    static_cast<ColumnVector<UInt32>*>(res_column->assume_mutable().get())
+                            ->get_data());
+        } else {
+            // neither DateTime nor DateTimeV2/DateV2, return null
+            if constexpr (std::is_same_v<DateType, DataTypeDateTime>) {
+                res_column = ColumnInt64::create(input_rows_count);
```
Comment (zhangstar333) [inline be/src/vec/functions/function_timestamp.cpp:655]: this check seems to have been done at the above

File: regression-test/suites/query_p0/sql_functions/datetime_functions/test_date_function.groovy (hunk 3 above)
Comment (lsy3993) [reply regression-test/suites/query_p0/sql_functions/datetime_functions/test_date_function.groovy:429]: done

File: be/src/vec/functions/function_timestamp.cpp (hunk 4 above)
Comment (lsy3993) [reply be/src/vec/functions/function_timestamp.cpp:582]: done

File: be/src/vec/functions/function_timestamp.cpp [hunk 6]
```diff
@@ -569,6 +569,164 @@ class FunctionUnixTimestamp : public IFunction {
...
+    bool use_default_implementation_for_nulls() const override { return true; }
+
+    size_t get_number_of_arguments() const override { return 1; }
+
+    bool is_variadic() const override { return true; }
+
+    // input DateTime and Date, return Date
+    // input DateTimeV2 and DateV2, return DateV2
+    DataTypePtr get_return_type_impl(const ColumnsWithTypeAndName& arguments) const override {
+        if constexpr (std::is_same_v<DateType, DataTypeDateTime>) {
```
[the remark this answers is not in the document]
Comment (lsy3993) [reply be/src/vec/functions/function_timestamp.cpp]: done

File: be/src/vec/functions/function_timestamp.cpp (hunk 5 above)
Comment (lsy3993) [reply be/src/vec/functions/function_timestamp.cpp]: you are right, I delete the 'else'

**Pull request closed without merging**